src/metabase/automagic_dashboards/rules.clj

```diff
@@ -276,79 +276,73 @@
-                                                (file->parent-dir f)
-                                                (file->table-type f)))))
-     (catch Exception e
-       (log/error (format "Error parsing %s:\n%s"
-                          (.getFileName f)
-                          (or (some-> e
-                                      ex-data
-                                      (select-keys [:error :value])
-                                      u/pprint-to-str)
-                              e)))
-       nil))))
-
-(defn load-rules
-  "Load and validate all rules in dir."
-  ([dir]
-   (with-resources fs
-     (load-rules fs dir)))
-  ([fs dir]
-   (when-let [dir (resource-path fs dir)]
-     (with-open [ds (Files/newDirectoryStream dir)]
-       (->> ds
-            (filter #(str/ends-with? (.toString ^Path %) ".yaml"))
-            (keep (partial load-rule fs))
-            doall)))))
-
-(defn -main
-  "Entry point for lein task `validate-automagic-dashboards`"
-  [& _]
-  (dorun (load-rules "tables"))
-  (dorun (load-rules "metrics"))
-  (dorun (load-rules "fields"))
-  (System/exit 0))
+(defn- load-rule
+  [f]
+  (try
+    (let [entity-type (file->entity-type f)]
+      (-> f
+          .toUri
+          slurp
+          yaml/parse-string
+          (assoc :rule entity-type)
+          (update :applies_to #(or % entity-type))
+          rules-validator))
+    (catch Exception e
+      (log/error (format "Error parsing %s:\n%s"
+                         (.getFileName f)
+                         (or (some-> e
+                                     ex-data
+                                     (select-keys [:error :value])
+                                     u/pprint-to-str)
+                             e)))
+      nil)))
+
+(defn- load-rule-dir
+  ([dir] (load-rule-dir dir [] {}))
+  ([dir path rules]
+   (with-open [ds (Files/newDirectoryStream dir)]
+     (reduce (fn [rules ^Path f]
+               (cond
+                 (Files/isDirectory f (into-array java.nio.file.LinkOption []))
+                 (load-rule-dir f (conj path (str (.getFileName f))) rules)
+
+                 (file->entity-type f)
+                 (assoc-in rules (concat path [(file->entity-type f) ::leaf]) (load-rule f))
+
+                 :else
+                 rules))
+             rules
+             ds))))
+
+(defmacro ^:private with-resource
+  [[identifier path] & body]
+  `(let [[jar# path#] (-> ~path .toString (str/split #"!" 2))]
+     (if path#
+       (with-open [^FileSystem fs# (-> jar#
+                                       java.net.URI/create
+                                       (FileSystems/newFileSystem (java.util.HashMap.)))]
+         (let [~identifier (.getPath fs# path# (into-array String []))]
+           ~@body))
+       (let [~identifier (.getPath (FileSystems/getDefault) (.getPath ~path) (into-array String []))]
+         ~@body))))
+
+(def ^:private rules
+  (with-resource [path (-> rules-dir io/resource .toURI)]
```

Comment:
This seems like the kind of thing that should be put behind a `delay` so that we don't pay this cost at startup and during compilation. Any reason why you're not doing that?

Reply:
No good reason. Fixed. Thanks.